File: swarm module/Blue_team/solaris-agent/vibecheck/core/supabase_client.py
"""
Supabase client for Project VibeCheck.

CRITICAL: The supabase Python SDK has NO async support.
All methods must be sync internally, wrapped in asyncio.get_event_loop().run_in_executor(None, sync_fn).

Used for:
- Scan status tracking
- Vulnerability storage
- Report generation
"""

import asyncio
import logging
from datetime import datetime, timezone
from typing import Any

from supabase import create_client, Client

from core.config import get_settings

logger = logging.getLogger(__name__)


class SupabaseClient:
    """
    Client for interacting with Supabase database.

    All methods are sync internally but exposed as async via run_in_executor.
    """

    def __init__(self, url: str | None = None, key: str | None = None) -> None:
        """
        Initialize Supabase client.

        Args:
            url: Supabase project URL (defaults to settings)
            key: Supabase anon/service key (defaults to settings)
        """
        settings = get_settings()
        self._url = url or settings.supabase_url
        self._key = key or settings.supabase_anon_key
        self._client: Client | None = None

    def _get_client(self) -> Client:
        """Get or create the Supabase client (sync)."""
        if self._client is None:
            logger.info(f"Connecting to Supabase at {self._url}")
            self._client = create_client(self._url, self._key)
            logger.info("Supabase connection established")
        return self._client

    # ==================== SCAN STATUS ====================

    def _get_scan_status_sync(self, scan_id: str) -> dict[str, Any] | None:
        """
        Sync method to get scan status.

        Args:
            scan_id: Unique scan identifier

        Returns:
            Scan record or None if not found
        """
        client = self._get_client()
        result = (
            client.table("scan_queue")
            .select("*")
            .eq("id", scan_id)
            .single()
            .execute()
        )
        return result.data if result else None

    async def get_scan_status(self, scan_id: str) -> dict[str, Any] | None:
        """
        Async wrapper to get scan status.

        Uses run_in_executor to avoid blocking the event loop.
        """
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            None,
            lambda: self._get_scan_status_sync(scan_id)
        )

    def _update_scan_status_sync(
        self,
        scan_id: str,
        status: str,
        progress: int = 0,
        error_message: str | None = None,
        current_stage: str | None = None,
        stage_output: dict | None = None,
    ) -> bool:
        """
        Sync method to update scan status.

        Args:
            scan_id: Unique scan identifier
            status: New status (pending, running, completed, failed)
            progress: Progress percentage (0-100)
            error_message: Optional error message
            current_stage: Optional current pipeline stage name
            stage_output: Optional dict with intermediate stage results

        Returns:
            True if update succeeded
        """
        client = self._get_client()
        
        update_data = {
            "status": status,
            "progress": progress,
        }
        
        if status == "running" and progress == 0:
            update_data["started_at"] = datetime.now(timezone.utc).isoformat()
        elif status in ("completed", "failed"):
            update_data["completed_at"] = datetime.now(timezone.utc).isoformat()
        
        if error_message:
            update_data["error_message"] = error_message
        
        if current_stage:
            update_data["current_stage"] = current_stage
        
        if stage_output:
            update_data["stage_output"] = stage_output
        
        result = (
            client.table("scan_queue")
            .update(update_data)
            .eq("id", scan_id)
            .execute()
        )
        
        return result is not None

    async def update_scan_status(
        self,
        scan_id: str,
        status: str,
        progress: int = 0,
        error_message: str | None = None,
        current_stage: str | None = None,
        stage_output: dict | None = None,
    ) -> bool:
        """
        Async wrapper to update scan status.
        """
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            None,
            lambda: self._update_scan_status_sync(scan_id, status, progress, error_message, current_stage, stage_output)
        )

    def _create_scan_sync(
        self,
        scan_id: str,
        repo_url: str,
        triggered_by: str = "unknown",
    ) -> dict[str, Any] | None:
        """
        Sync method to create a new scan record.

        Args:
            scan_id: Unique scan identifier
            repo_url: Repository URL
            triggered_by: User who triggered the scan

        Returns:
            Created record or None
        """
        client = self._get_client()
        
        result = (
            client.table("scan_queue")
            .insert({
                "id": scan_id,
                "repo_url": repo_url,
                "triggered_by": triggered_by,
                "status": "pending",
                "progress": 0,
            })
            .execute()
        )
        
        return result.data[0] if result and result.data else None

    async def create_scan(
        self,
        scan_id: str,
        repo_url: str,
        triggered_by: str = "unknown",
    ) -> dict[str, Any] | None:
        """
        Async wrapper to create a new scan record.
        """
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            None,
            lambda: self._create_scan_sync(scan_id, repo_url, triggered_by)
        )

    # ==================== VULNERABILITIES ====================

    def _get_vulnerabilities_sync(self, scan_id: str) -> list[dict[str, Any]]:
        """
        Sync method to get vulnerabilities for a scan.

        Args:
            scan_id: Unique scan identifier

        Returns:
            List of vulnerability records
        """
        client = self._get_client()
        result = (
            client.table("vulnerabilities")
            .select("*")
            .eq("scan_id", scan_id)
            .execute()
        )
        return result.data if result else []

    async def get_vulnerabilities(self, scan_id: str) -> list[dict[str, Any]]:
        """
        Async wrapper to get vulnerabilities.
        """
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            None,
            lambda: self._get_vulnerabilities_sync(scan_id)
        )

    def _insert_vulnerability_sync(
        self,
        scan_id: str,
        vuln: dict[str, Any],
    ) -> dict[str, Any] | None:
        """
        Sync method to insert a vulnerability.

        Args:
            scan_id: Unique scan identifier
            vuln: Vulnerability data

        Returns:
            Created record or None
        """
        client = self._get_client()
        
        # Ensure scan_id is included
        vuln_data = {**vuln, "scan_id": scan_id}
        
        result = (
            client.table("vulnerabilities")
            .insert(vuln_data)
            .execute()
        )
        
        return result.data[0] if result and result.data else None

    async def insert_vulnerability(
        self,
        scan_id: str,
        vuln: dict[str, Any],
    ) -> dict[str, Any] | None:
        """
        Async wrapper to insert a vulnerability.
        """
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            None,
            lambda: self._insert_vulnerability_sync(scan_id, vuln)
        )

    def _insert_vulnerabilities_batch_sync(
        self,
        scan_id: str,
        vulns: list[dict[str, Any]],
    ) -> list[dict[str, Any]]:
        """
        Sync method to insert multiple vulnerabilities.

        Args:
            scan_id: Unique scan identifier
            vulns: List of vulnerability data

        Returns:
            List of created records
        """
        if not vulns:
            logger.warning("No vulnerabilities to insert - vulns list is empty")
            return []
        
        client = self._get_client()
        
        # First, verify the scan_id exists in scan_queue (FK constraint)
        try:
            scan_check = (
                client.table("scan_queue")
                .select("id")
                .eq("id", scan_id)
                .execute()
            )
            if not scan_check.data:
                logger.error(f"Scan ID {scan_id} does not exist in scan_queue table - FK constraint will fail")
                raise ValueError(f"Scan ID {scan_id} not found in scan_queue table")
            logger.info(f"Verified scan_id {scan_id} exists in scan_queue")
        except Exception as e:
            logger.error(f"Failed to verify scan_id: {e}")
            raise
        
        # Add scan_id to each vulnerability
        vulns_data = [{**v, "scan_id": scan_id} for v in vulns]
        
        logger.info(f"Inserting {len(vulns_data)} vulnerabilities into Supabase for scan {scan_id}")
        logger.debug(f"Sample vulnerability data: {vulns_data[0] if vulns_data else 'none'}")
        
        try:
            # Use upsert to handle unique constraint on (scan_id, file_path, line_start)
            # This prevents race conditions where duplicate vulns are inserted
            result = (
                client.table("vulnerabilities")
                .upsert(vulns_data, on_conflict="scan_id,file_path,line_start")
                .execute()
            )
            
            if result and result.data:
                logger.info(f"Successfully inserted {len(result.data)} vulnerabilities")
                return result.data
            else:
                logger.warning(f"Insert returned no data. Result: {result}")
                return []
        except Exception as e:
            logger.error(f"Failed to insert vulnerabilities: {e}", exc_info=True)
            raise

    async def insert_vulnerabilities_batch(
        self,
        scan_id: str,
        vulns: list[dict[str, Any]],
    ) -> list[dict[str, Any]]:
        """
        Async wrapper to insert multiple vulnerabilities.
        """
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            None,
            lambda: self._insert_vulnerabilities_batch_sync(scan_id, vulns)
        )

    # ==================== REPORTS ====================

    def _get_report_sync(self, scan_id: str) -> dict[str, Any] | None:
        """
        Sync method to get a full report for a scan.

        Args:
            scan_id: Unique scan identifier

        Returns:
            Report data with scan info and vulnerabilities
        """
        client = self._get_client()
        
        # Get scan info
        scan_result = (
            client.table("scan_queue")
            .select("*")
            .eq("id", scan_id)
            .single()
            .execute()
        )
        
        if not scan_result or not scan_result.data:
            return None
        
        # Get vulnerabilities
        vulns_result = (
            client.table("vulnerabilities")
            .select("*")
            .eq("scan_id", scan_id)
            .execute()
        )
        
        return {
            "scan": scan_result.data,
            "vulnerabilities": vulns_result.data if vulns_result else [],
        }

    async def get_report(self, scan_id: str) -> dict[str, Any] | None:
        """
        Async wrapper to get a full report.
        """
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            None,
            lambda: self._get_report_sync(scan_id)
        )

    # ==================== LIST SCANS ====================

    def _list_scans_sync(
        self,
        status: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        Sync method to list scans.

        Args:
            status: Optional status filter
            limit: Maximum number of results
            offset: Offset for pagination

        Returns:
            Dict with "scans" list and "total" count
        """
        client = self._get_client()
        
        # Build query for paginated results
        query = client.table("scan_queue").select("*")
        
        if status:
            query = query.eq("status", status)
        
        result = (
            query.order("created_at", desc=True)
            .limit(limit)
            .offset(offset)
            .execute()
        )
        
        scans = result.data if result else []
        
        # Get total count (separate query)
        count_query = client.table("scan_queue").select("*", count="exact")
        
        if status:
            count_query = count_query.eq("status", status)
        
        count_result = count_query.execute()
        total = count_result.count if count_result else len(scans)
        
        return {
            "scans": scans,
            "total": total,
        }

    async def list_scans(
        self,
        status: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        Async wrapper to list scans.
        
        Returns:
            Dict with "scans" list and "total" count
        """
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            None,
            lambda: self._list_scans_sync(status, limit, offset)
        )


# Singleton instance
_client: SupabaseClient | None = None


def get_supabase_client() -> SupabaseClient:
    """Get the Supabase client singleton."""
    global _client
    if _client is None:
        _client = SupabaseClient()
    return _client

File: vibecheck-refactored/src/db/clients/supabase-client.ts
import { createClient, SupabaseClient as SupabaseSDK } from "@supabase/supabase-js";

interface SupabaseClientOptions {
  url: string;
  serviceKey: string;
}

export class SupabaseClient {
  private client: SupabaseSDK | null = null;
  private _isConnected = false;

  constructor(private options: SupabaseClientOptions) {}

  get isConnected(): boolean {
    return this._isConnected;
  }

  async connect(): Promise<void> {
    this.client = createClient(
      this.options.url,
      this.options.serviceKey,
      {
        auth: {
          persistSession: false,
          autoRefreshToken: false,
        },
      }
    );
    this._isConnected = true;
  }

  async disconnect(): Promise<void> {
    this.client = null;
    this._isConnected = false;
  }

  from(table: string) {
    if (!this.client) throw new Error("Client not connected");
    return this.client.from(table);
  }

  async insert<T = Record<string, unknown>>(
    table: string,
    data: Partial<T>
  ): Promise<T | null> {
    if (!this.client) throw new Error("Client not connected");

    const { data: result, error } = await this.client.from(table).insert(data as Record<string, unknown>).select().single();

    if (error) {
      console.error("Supabase insert error:", error);
      return null;
    }

    return result as T;
  }

  async insertMany<T = Record<string, unknown>>(
    table: string,
    data: Partial<T>[]
  ): Promise<T[] | null> {
    if (!this.client) throw new Error("Client not connected");

    const { data: result, error } = await this.client.from(table).insert(data as Record<string, unknown>[]).select();

    if (error) {
      console.error("Supabase insertMany error:", error);
      return null;
    }

    return result as T[];
  }

  update<T = Record<string, unknown>>(
    table: string,
    data: Partial<T>
  ) {
    if (!this.client) throw new Error("Client not connected");
    return this.client.from(table).update(data as Record<string, unknown>);
  }

  delete(table: string) {
    if (!this.client) throw new Error("Client not connected");
    return this.client.from(table).delete();
  }

  async query<T = Record<string, unknown>>(
    table: string,
    options?: {
      select?: string;
      where?: Record<string, unknown>;
      limit?: number;
      orderBy?: { column: string; ascending?: boolean };
    }
  ): Promise<T[]> {
    if (!this.client) throw new Error("Client not connected");

    let query = this.client.from(table).select(options?.select || "*");

    if (options?.where) {
      for (const [key, value] of Object.entries(options.where)) {
        query = query.eq(key, value);
      }
    }

    if (options?.limit) {
      query = query.limit(options.limit);
    }

    if (options?.orderBy) {
      query = query.order(options.orderBy.column, {
        ascending: options.orderBy.ascending ?? true,
      });
    }

    const { data, error } = await query;

    if (error) {
      console.error("Supabase query error:", error);
      return [];
    }

    return data as T[];
  }

  async listSwarmMissions(limit = 20, offset = 0) {
    return this.query("swarm_missions", {
      orderBy: { column: "created_at", ascending: false },
      limit,
    });
  }

  async createSwarmMission(missionId: string, target: string, objective: string, mode: string, maxIterations: number, scanId?: string) {
    return this.insert("swarm_missions", {
      id: missionId,
      target,
      objective,
      mode,
      max_iterations: maxIterations,
      scan_id: scanId,
      status: "pending",
      progress: 0,
      iteration: 0,
    });
  }

  async getSwarmMission(missionId: string) {
    if (!this.client) throw new Error("Client not connected");
    const { data, error } = await this.client.from("swarm_missions").select("*").eq("id", missionId).single();
    if (error) return null;
    return data;
  }

  async updateSwarmMission(missionId: string, updates: Record<string, unknown>) {
    if (!this.client) throw new Error("Client not connected");
    const { data, error } = await this.client.from("swarm_missions").update(updates).eq("id", missionId).select().single();
    if (error) return null;
    return data;
  }

  async getSwarmAgentStates(missionId: string) {
    return this.query("swarm_agent_states", {
      where: { mission_id: missionId },
      orderBy: { column: "created_at", ascending: true },
    });
  }

  async createSwarmAgentState(
    missionId: string,
    agentId: string,
    agentName: string,
    agentTeam: string,
    status: string,
    iter: string,
    task: string
  ) {
    return this.insert("swarm_agent_states", {
      mission_id: missionId,
      agent_id: agentId,
      agent_name: agentName,
      agent_team: agentTeam,
      status,
      iter,
      task,
      last_updated: new Date().toISOString(),
    });
  }

  async getSwarmAgentEvents(missionId: string, limit = 100, agentName?: string) {
    if (!this.client) throw new Error("Client not connected");
    let q = this.client.from("swarm_agent_events").select("*").eq("mission_id", missionId).order("created_at", { ascending: false }).limit(limit);
    if (agentName) q = q.eq("agent_name", agentName);
    const { data, error } = await q;
    if (error) return [];
    return data;
  }

  async getSwarmFindings(missionId: string) {
    return this.query("swarm_findings", { where: { mission_id: missionId } });
  }

  async getSwarmEvents(missionId: string, limit = 100, eventType?: string, agentName?: string, iteration?: number) {
    if (!this.client) throw new Error("Client not connected");
    let q = this.client.from("swarm_events").select("*").eq("mission_id", missionId).order("created_at", { ascending: false }).limit(limit);
    if (eventType) q = q.eq("event_type", eventType);
    if (agentName) q = q.eq("agent_name", agentName);
    if (iteration !== undefined) q = q.eq("iteration", iteration);
    const { data, error } = await q;
    if (error) return [];
    return data;
  }

  async getMissionTimeline(missionId: string) {
    if (!this.client) throw new Error("Client not connected");
    const { data, error } = await this.client.from("mission_timeline").select("*").eq("mission_id", missionId).order("created_at", { ascending: true });
    if (error) return [];
    return data;
  }

  async getMissionStatistics(missionId: string) {
    if (!this.client) throw new Error("Client not connected");
    const { data, error } = await this.client.from("mission_statistics").select("*").eq("mission_id", missionId).single();
    if (error) return null;
    return data;
  }

  async getScanStatus(scanId: string) {
    if (!this.client) throw new Error("Client not connected");
    const { data, error } = await this.client.from("scan_queue").select("*").eq("id", scanId).single();
    if (error) return null;
    return data;
  }

  async updateScanStatus(
    scanId: string,
    status: string,
    progress: number = 0,
    errorMessage?: string,
    currentStage?: string,
    stageOutput?: Record<string, unknown>
  ): Promise<boolean> {
    if (!this.client) throw new Error("Client not connected");

    const updateData: Record<string, unknown> = {
      status,
      progress,
    };

    if (status === "running" && progress === 0) {
      updateData.started_at = new Date().toISOString();
    } else if (status === "completed" || status === "failed") {
      updateData.completed_at = new Date().toISOString();
    }

    if (errorMessage) {
      updateData.error_message = errorMessage;
    }

    if (currentStage) {
      updateData.current_stage = currentStage;
    }

    if (stageOutput) {
      updateData.stage_output = stageOutput;
    }

    const { error } = await this.client.from("scan_queue").update(updateData).eq("id", scanId);
    return !error;
  }

  async createScan(scanId: string, repoUrl: string, triggeredBy: string = "unknown") {
    if (!this.client) throw new Error("Client not connected");
    const { data, error } = await this.client.from("scan_queue").insert({
      id: scanId,
      repo_url: repoUrl,
      triggered_by: triggeredBy,
      status: "pending",
      progress: 0,
    }).select().single();

    if (error) return null;
    return data;
  }

  async getVulnerabilities(scanId: string) {
    if (!this.client) throw new Error("Client not connected");
    const { data, error } = await this.client.from("vulnerabilities").select("*").eq("scan_id", scanId);
    if (error) return [];
    return data || [];
  }

  async insertVulnerability(scanId: string, vuln: Record<string, unknown>) {
    if (!this.client) throw new Error("Client not connected");
    const vulnData = { ...vuln, scan_id: scanId };
    const { data, error } = await this.client.from("vulnerabilities").insert(vulnData).select().single();
    if (error) return null;
    return data;
  }

  async insertVulnerabilitiesBatch(scanId: string, vulns: Record<string, unknown>[]) {
    if (!this.client) throw new Error("Client not connected");
    if (!vulns.length) return [];

    const scanCheck = await this.client.from("scan_queue").select("id").eq("id", scanId).single();
    if (scanCheck.error || !scanCheck.data) {
      throw new Error(`Scan ID ${scanId} not found in scan_queue table`);
    }

    const vulnsData = vulns.map(v => ({ ...v, scan_id: scanId }));
    const { data, error } = await this.client
      .from("vulnerabilities")
      .upsert(vulnsData, { onConflict: "scan_id,file_path,line_start" })
      .select();

    if (error) throw error;
    return data || [];
  }

  async getReport(scanId: string) {
    if (!this.client) throw new Error("Client not connected");

    const { data: scan, error: scanError } = await this.client
      .from("scan_queue")
      .select("*")
      .eq("id", scanId)
      .single();

    if (scanError || !scan) return null;

    const { data: vulns, error: vulnsError } = await this.client
      .from("vulnerabilities")
      .select("*")
      .eq("scan_id", scanId);

    return {
      scan,
      vulnerabilities: vulns || [],
    };
  }

  async listScans(status?: string, limit: number = 50, offset: number = 0) {
    if (!this.client) throw new Error("Client not connected");

    let query = this.client.from("scan_queue").select("*", { count: "exact" });

    if (status) {
      query = query.eq("status", status);
    }

    const { data: scans, error, count } = await query
      .order("created_at", { ascending: false })
      .range(offset, offset + limit - 1);

    if (error) {
      return { scans: [], total: 0 };
    }

    return {
      scans: scans || [],
      total: count ?? scans?.length ?? 0,
    };
  }

  async getSwarmExploitAttempts(missionId: string, limit = 500, exploitType?: string, success?: boolean) {
    if (!this.client) throw new Error("Client not connected");
    let q = this.client.from("swarm_exploit_attempts").select("*").eq("mission_id", missionId).order("created_at", { ascending: false }).limit(limit);
    if (exploitType) q = q.eq("exploit_type", exploitType);
    if (success !== undefined) q = q.eq("success", success);
    const { data, error } = await q;
    if (error) return [];
    return data;
  }
}

export const supabaseClient = new SupabaseClient({
  url: process.env.SUPABASE_URL || "",
  serviceKey: process.env.SUPABASE_SERVICE_KEY || "",
});